tests/storage/restore.go

```diff
@@ -1493,6 +1493,133 @@ var _ = SIGDescribe("VirtualMachineRestore Tests", func() {
+					cloneRoleBinding = roleBinding
+				})
+
+				AfterEach(func() {
+					if sourceDV != nil {
+						err := virtClient.CdiClient().CdiV1beta1().DataVolumes(sourceDV.Namespace).Delete(context.TODO(), sourceDV.Name, metav1.DeleteOptions{})
+						Expect(err).ToNot(HaveOccurred())
+					}
+
+					if cloneRole != nil {
+						err := virtClient.RbacV1().Roles(cloneRole.Namespace).Delete(context.TODO(), cloneRole.Name, metav1.DeleteOptions{})
+						Expect(err).ToNot(HaveOccurred())
+					}
+					if cloneRoleBinding != nil {
+						err = virtClient.RbacV1().RoleBindings(cloneRoleBinding.Namespace).Delete(context.TODO(), cloneRoleBinding.Name, metav1.DeleteOptions{})
+						Expect(err).ToNot(HaveOccurred())
+					}
+				})
+
+				checkNoCloneAnnotations := func(vm *v1.VirtualMachine, shouldExist bool) {
+					pvcName := ""
+					for _, v := range vm.Spec.Template.Spec.Volumes {
+						if v.DataVolume != nil {
+							Expect(pvcName).Should(Equal(""))
+							pvcName = v.DataVolume.Name
+						} else if v.PersistentVolumeClaim != nil {
+							Expect(pvcName).Should(Equal(""))
+							pvcName = v.PersistentVolumeClaim.ClaimName
+						}
+					}
+					pvc, err := virtClient.CoreV1().PersistentVolumeClaims(vm.Namespace).Get(context.TODO(), pvcName, metav1.GetOptions{})
+					Expect(err).ToNot(HaveOccurred())
+					for _, a := range []string{"k8s.io/CloneRequest", "k8s.io/CloneOf"} {
+						_, ok := pvc.Annotations[a]
+						Expect(ok).Should(Equal(shouldExist))
+					}
+				}
+
+				DescribeTable("should restore a vm that boots from a network cloned datavolumetemplate", func(restoreToNewVM bool) {
+					vm, vmi = createAndStartVM(tests.NewRandomVMWithDataVolumeCloneSourceAndUserData(
+						sourceDV.Namespace,
+						sourceDV.Name,
+						util.NamespaceTestDefault,
+						bashHelloScript,
+						snapshotStorageClass,
+						corev1.ReadWriteOnce,
+					))
+
+					checkNoCloneAnnotations(vm, true)
+					doRestore("", console.LoginToCirros, offlineSnaphot, getTargetVMName(restoreToNewVM, newVmName))
+					checkNoCloneAnnotations(getTargetVM(restoreToNewVM), false)
+				},
+					Entry("to the same VM", false),
+					Entry("to a new VM", true),
+				)
+
+				DescribeTable("should restore a vm that boots from a network cloned datavolume (not template)", func(restoreToNewVM bool) {
+					vm = tests.NewRandomVMWithDataVolumeCloneSourceAndUserData(
+						sourceDV.Namespace,
+						sourceDV.Name,
+						util.NamespaceTestDefault,
+						bashHelloScript,
+						snapshotStorageClass,
+						corev1.ReadWriteOnce,
+					)
+
+					dvt := &vm.Spec.DataVolumeTemplates[0]
+
+					dv := &cdiv1.DataVolume{}
+					dv.ObjectMeta = *dvt.ObjectMeta.DeepCopy()
+					dv.Spec = *dvt.Spec.DeepCopy()
+
+					vm.Spec.DataVolumeTemplates = nil
+
+					dv, err = virtClient.CdiClient().CdiV1beta1().DataVolumes(vm.Namespace).Create(context.Background(), dv, metav1.CreateOptions{})
+					Expect(err).ToNot(HaveOccurred())
+					dv = waitDVReady(dv)
+					defer func() {
```

Comment:
shouldn't this defer be before the waitDVReady? if the ready condition fails it will not clean the datavolume that was already created?

Reply:
yup